services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy

```diff
@@ -0,0 +1,204 @@
+        given: "Two flows with #flowConfig configuration"
+        def (src, dest) = switchSelectMethod()
+        def flowToBreak = northbound.addFlow(flowHelper.randomFlow(src, dest))
+        def intactFlow = northbound.addFlow(flowHelper.randomFlow(src, dest))
+
+        waitForFlowSetup(flowToBreak.id)
+        waitForFlowSetup(intactFlow.id)
+
+        and: "Both flows have the same switches in path"
+        def damagedFlowSwitches = pathHelper.getInvolvedSwitches(flowToBreak.id)*.dpId
+        def intactFlowSwitches = pathHelper.getInvolvedSwitches(intactFlow.id)*.dpId
+        assert damagedFlowSwitches.equals(intactFlowSwitches)
+
+        when: "#flowType flow rule from first flow on #ruleDelete switch gets deleted"
+        Flow damagedFlow = flowType == "forward" ? database.getFlow(flowToBreak.id).left :
+                database.getFlow(flowToBreak.id).right
+        SwitchId damagedSwitch = damagedFlowSwitches[item]
+        northbound.deleteSwitchRules(damagedSwitch, damagedFlow.cookie)
+
+        then: "Intact flow should be validated successfully"
+        northbound.validateFlow(intactFlow.id).every { isFlowValid(it) }
+
+        and: "Damaged #flowType flow validation should fail, while other direction should be validated successfully"
+        def validationResult = northbound.validateFlow(flowToBreak.id)
+        validationResult.findAll {isFlowValid(it)} .size() == 1
+        def invalidFlow = validationResult.findAll {!isFlowValid(it)}
+        invalidFlow.size() == 1
+
+        and: "Flow rule discrepancy should contain dpID of the affected switch and cookie of the damaged flow"
+        def rules = findRulesDiscrepancies(invalidFlow[0])
+        rules.size() == 1
+        rules[damagedSwitch.toString()] == damagedFlow.cookie.toString()
+
+        and: "Validation of non-affected flow should succeed"
+        northbound.validateFlow(intactFlow.id).every { isFlowValid(it) }
+
+        and: "Affected switch should have one missing rule with the same cookie as the damaged flow"
+        def switchValidationResult = northbound.validateSwitchRules(damagedSwitch)
+        switchValidationResult.missingRules.size() == 1
+        switchValidationResult.missingRules[0] == damagedFlow.cookie
+
+        and: "There should be no excess rules on the affected switch"
+        switchValidationResult.excessRules.size() == 0
+
+        and: "Validation of non-affected switches (if any) should succeed"
+        if (damagedFlowSwitches.size() > 1) {
+            def nonAffectedSwitches = damagedFlowSwitches.findAll { it != damagedFlowSwitches[item] }
+            nonAffectedSwitches.every { sw -> northbound.validateSwitchRules(sw).missingRules.size() == 0 }
+            nonAffectedSwitches.every { sw -> northbound.validateSwitchRules(sw).excessRules.size() == 0 }
+        }
+
+        and: "Delete the flows"
+        [flowToBreak?.id, intactFlow?.id].each { deleteFlow(it) }
```

Comment:
your safe navigation here makes very little sense. If, for example, intact flow will be null, then you'll loop over
`[flowtoBreakId, null]. each { deleteFlow(it) }` and will still pass null to the deleteFlow method which will anyway result in NPE.  
Also, your safe navigation makes no sense, because previously in the test you already accessed `flowToBreak.id` without safe navigation. Since this is not a `cleanup` block, this code will only get executed if previous code worked, so..

Reply:
Makes sense, thanks